addons/mail/static/src/model/record.js

```diff
@@ -409,20 +416,75 @@ export class Record {
         return !this.in(collection);
     }
 
-    toData() {
+    /**
+     * Converts the current record and its related data into Store insert-able data.
+     * @param {Array<string> | { depth: boolean }} options Configuration options or an array of field names.
+     * @returns {Object} A data object grouped by model names.
+     */
+    toData(options = { depth: false }) {
+        const prefix = this._getPyModelName();
+        const ongoing = {
+            seenRecords: new Set(),
+            storeData: {},
+            depth: options.depth,
+            fields: undefined,
+        };
+        if (Array.isArray(options)) {
+            ongoing.fields = options.map((field) => `${prefix}.${field}`);
+        }
+        this._toData(ongoing, prefix);
+        return ongoing.storeData;
+    }
+
+    _cleanupData(data) {
+        const fieldsToDelete = [
+            "_",
+            "_fieldsValue",
+            "_proxy",
+            "_proxyInternal",
+            "_raw",
+            "env",
+            "Model",
+        ];
+        fieldsToDelete.forEach((field) => delete data[field]);
+    }
+
+    _getPyModelName() {
+        return this.Model.getName();
+    }
+
+    /**
+     * @param {Ongoing} ongoing The ongoing data conversion state.
+     * @param {string} [prefix] The prefix for the current field (used for nested fields).
+     */
+    _toData(ongoing, prefix = undefined) {
+        if (ongoing.seenRecords.has(this.localId)) {
+            return;
+        }
```

Comment:
I'm not fully sure, but couldn't that prevent from following deeper relations if the seen record was added without said relations but the new request is asking to add them?

Reply:
So something like `message.toData("author", "lastReaction.author.group_ids"`)? Could indeed prevent retrieving the `group_ids` data in that case. Should take into account `ongoing.fields` with likely some tracking of whose data have been retrieved already.

Reply:
Either that or just add as many times as it appears, it's going to overwrite the same data anyway no?